# Fase 1: Seleccion y reconocimiento de datos

In [2]:
#libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pathlib as pl
from libmac import listar_archivos, etiquetar_unirdatosdeclase

## 1. Dataset del autor (flujos presentados)

In [3]:
#import data
dfa_normal = pd.read_csv('InSDN_DatasetCSV/Normal_data.csv')
dfa_ataques = pd.read_csv('InSDN_DatasetCSV/OVS.csv')

### 1.1 Tráfico normal

In [4]:
#show features db normal
dfa_normal.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68424 entries, 0 to 68423
Data columns (total 84 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Flow ID            68424 non-null  object 
 1   Src IP             68424 non-null  object 
 2   Src Port           68424 non-null  int64  
 3   Dst IP             68424 non-null  object 
 4   Dst Port           68424 non-null  int64  
 5   Protocol           68424 non-null  int64  
 6   Timestamp          68424 non-null  object 
 7   Flow Duration      68424 non-null  int64  
 8   Tot Fwd Pkts       68424 non-null  int64  
 9   Tot Bwd Pkts       68424 non-null  int64  
 10  TotLen Fwd Pkts    68424 non-null  float64
 11  TotLen Bwd Pkts    68424 non-null  float64
 12  Fwd Pkt Len Max    68424 non-null  int64  
 13  Fwd Pkt Len Min    68424 non-null  int64  
 14  Fwd Pkt Len Mean   68424 non-null  float64
 15  Fwd Pkt Len Std    68424 non-null  float64
 16  Bwd Pkt Len Max    684

In [5]:
#show statics of db normal
dfa_normal.describe()

Src Port      Dst Port      Protocol  Flow Duration  Tot Fwd Pkts  \
count  68424.000000  68424.000000  68424.000000   6.842400e+04  68424.000000   
mean   37361.032459   9733.738557     11.418947   1.362636e+07     11.807217   
std    20305.851419  19352.308798      5.551915   3.360107e+07    112.712183   
min        0.000000      0.000000      0.000000  -1.540000e+02      0.000000   
25%    34213.500000     53.000000      6.000000   3.134000e+03      1.000000   
50%    43383.000000    443.000000      6.000000   4.799000e+03      1.000000   
75%    52105.250000    443.000000     17.000000   7.122470e+05      6.000000   
max    65513.000000  60998.000000     17.000000   1.200000e+08  16928.000000   

       Tot Bwd Pkts  TotLen Fwd Pkts  TotLen Bwd Pkts  Fwd Pkt Len Max  \
count  68424.000000     6.842400e+04     6.842400e+04     68424.000000   
mean      18.592628     3.354497e+03     4.128116e+04       425.863644   
std      235.949827     1.560998e+05     7.680108e+05      1425.400448   
min        1.000000     0.000000e+00     0.000000e+00         0.000000   
25%        2.000000     3.000000e+01     1.370000e+02        30.000000   
50%        3.000000     3.600000e+01     5.020000e+02        36.000000   
75%        5.000000     5.990000e+02     7.620000e+02       249.000000   
max    34094.000000     3.160000e+07     1.070000e+08     64239.000000   

       Fwd Pkt Len Min  ...  Fwd Act Data Pkts  Fwd Seg Size Min  \
count     68424.000000  ...       68424.000000           68424.0   
mean         22.297469  ...           3.891266               0.0   
std          67.228213  ...          57.082340               0.0   
min           0.000000  ...           0.000000               0.0   
25%           0.000000  ...           1.000000               0.0   
50%           0.000000  ...           1.000000               0.0   
75%          34.000000  ...           3.000000               0.0   
max        3900.000000  ...       11180.000000               0.0   

        Active Mean    Active Std    Active Max    Active Min     Idle Mean  \
count  6.842400e+04  6.842400e+04  6.842400e+04  6.842400e+04  6.842400e+04   
mean   1.887697e+05  1.434025e+05  3.980718e+05  8.563484e+04  4.508770e+06   
std    1.373001e+06  1.105629e+06  2.467808e+06  1.027435e+06  1.314630e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max    1.040000e+08  6.830000e+07  1.040000e+08  1.040000e+08  1.190000e+08   

           Idle Std      Idle Max      Idle Min  
count  6.842400e+04  6.842400e+04  6.842400e+04  
mean   6.041184e+05  5.066966e+06  3.975015e+06  
std    3.486207e+06  1.448106e+07  1.246315e+07  
min    0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  0.000000e+00  
75%    0.000000e+00  0.000000e+00  0.000000e+00  
max    7.300000e+07  1.190000e+08  1.190000e+08  

[8 rows x 79 columns]

In [6]:
#encontrar tipos de protocolos
dfa_normal.Protocol.unique()

array([ 6,  0, 17], dtype=int64)

In [7]:
#tipo de trafico
dfa_normal.Label.unique()

array(['Normal'], dtype=object)

**Descripción subconjunto de tráfico normal:**<br>
El subconjunto de tráfico normal cuenta con un total de 84 caracteríticas y 68424 muestras. Los registros de flujo analizados contiene tres tipos de datos: object, float e int. Las caracteristicas de tipo object son: 
- Flow ID (contiene la IP de origen y destino, puerto de origen y destino, y protocolo L4).
- Src IP (direccion ip de origen).
- Dst IP (direccion ip de destino). 
- Timestamp: marca de tiempo.
- Label: etiqueta.

La caracteristica Protocolo es de tipo Int y contiene 3 valores; HOPOPT, UDP, TCP, además estos datos ya se encuentran enumerados.
También, existen conjuntos de atributos estadísticos referentes a los flujos de tráficos, intervalos de tiempo, bytes, paquetes, entre otros. Estos últimos se analizarán con más detalle en la siguiente fase.

### 1.2 Tráfico de ataque

In [10]:
#mostrar caracteristicas de los datos de ataques
dfa_ataques.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138722 entries, 0 to 138721
Data columns (total 84 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Flow ID            138722 non-null  object 
 1   Src IP             138722 non-null  object 
 2   Src Port           138722 non-null  int64  
 3   Dst IP             138722 non-null  object 
 4   Dst Port           138722 non-null  int64  
 5   Protocol           138722 non-null  int64  
 6   Timestamp          138722 non-null  object 
 7   Flow Duration      138722 non-null  int64  
 8   Tot Fwd Pkts       138722 non-null  int64  
 9   Tot Bwd Pkts       138722 non-null  int64  
 10  TotLen Fwd Pkts    138722 non-null  int64  
 11  TotLen Bwd Pkts    138722 non-null  int64  
 12  Fwd Pkt Len Max    138722 non-null  int64  
 13  Fwd Pkt Len Min    138722 non-null  int64  
 14  Fwd Pkt Len Mean   138722 non-null  float64
 15  Fwd Pkt Len Std    138722 non-null  float64
 16  Bw

In [9]:
#mostrar estadisticas del trafico de ataques
dfa_ataques.describe()

Src Port      Dst Port      Protocol  Flow Duration  Tot Fwd Pkts  \
count  68424.000000  68424.000000  68424.000000   6.842400e+04  68424.000000   
mean   37361.032459   9733.738557     11.418947   1.362636e+07     11.807217   
std    20305.851419  19352.308798      5.551915   3.360107e+07    112.712183   
min        0.000000      0.000000      0.000000  -1.540000e+02      0.000000   
25%    34213.500000     53.000000      6.000000   3.134000e+03      1.000000   
50%    43383.000000    443.000000      6.000000   4.799000e+03      1.000000   
75%    52105.250000    443.000000     17.000000   7.122470e+05      6.000000   
max    65513.000000  60998.000000     17.000000   1.200000e+08  16928.000000   

       Tot Bwd Pkts  TotLen Fwd Pkts  TotLen Bwd Pkts  Fwd Pkt Len Max  \
count  68424.000000     6.842400e+04     6.842400e+04     68424.000000   
mean      18.592628     3.354497e+03     4.128116e+04       425.863644   
std      235.949827     1.560998e+05     7.680108e+05      1425.400448   
min        1.000000     0.000000e+00     0.000000e+00         0.000000   
25%        2.000000     3.000000e+01     1.370000e+02        30.000000   
50%        3.000000     3.600000e+01     5.020000e+02        36.000000   
75%        5.000000     5.990000e+02     7.620000e+02       249.000000   
max    34094.000000     3.160000e+07     1.070000e+08     64239.000000   

       Fwd Pkt Len Min  ...  Fwd Act Data Pkts  Fwd Seg Size Min  \
count     68424.000000  ...       68424.000000           68424.0   
mean         22.297469  ...           3.891266               0.0   
std          67.228213  ...          57.082340               0.0   
min           0.000000  ...           0.000000               0.0   
25%           0.000000  ...           1.000000               0.0   
50%           0.000000  ...           1.000000               0.0   
75%          34.000000  ...           3.000000               0.0   
max        3900.000000  ...       11180.000000               0.0   

        Active Mean    Active Std    Active Max    Active Min     Idle Mean  \
count  6.842400e+04  6.842400e+04  6.842400e+04  6.842400e+04  6.842400e+04   
mean   1.887697e+05  1.434025e+05  3.980718e+05  8.563484e+04  4.508770e+06   
std    1.373001e+06  1.105629e+06  2.467808e+06  1.027435e+06  1.314630e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max    1.040000e+08  6.830000e+07  1.040000e+08  1.040000e+08  1.190000e+08   

           Idle Std      Idle Max      Idle Min  
count  6.842400e+04  6.842400e+04  6.842400e+04  
mean   6.041184e+05  5.066966e+06  3.975015e+06  
std    3.486207e+06  1.448106e+07  1.246315e+07  
min    0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  0.000000e+00  
75%    0.000000e+00  0.000000e+00  0.000000e+00  
max    7.300000e+07  1.190000e+08  1.190000e+08  

[8 rows x 79 columns]

In [11]:
#encontrar tipos de protocolos
dfa_ataques.Protocol.unique()

array([ 6,  0, 17], dtype=int64)

In [12]:
# tipos de ataques
dfa_ataques.Label.unique()

array(['BFA', 'DDoS ', 'DoS', 'Probe', 'Web-Attack', 'BOTNET'],
      dtype=object)

**Descripción subconjunto de tráfico de ataque:**<br>
El subconjunto de tráfico de ataque cuenta con 84 caracteríticas y 138722 muestras. Los tipos de ataque son: fuerza bruta (BFA), DDoS, Dos, Probe, Web y botnet, cuyas muestras están desequilibradas. Así mismo, los registros de flujo contiene tres tipos de datos: object, float e int. Las caracteristicas de tipo object siguen siendo las mismas del anterior (normal). Se repiten los mismos protocolos codificados de L4 HOPOPT, UDP y TCP.
También, existen los mismos conjuntos de atributos estadísticos referentes a los flujos de tráficos, intervalos de tiempo, bytes, paquetes, etc, los cuales serán analizados con más detalle en la siguiente fase.

### 1.3 Total de datos

In [13]:
#concatenar dataset
df_unido = pd.concat([dfa_normal,dfa_ataques], axis=0)
#dimension del conjunto total 
df_unido.shape

(207146, 84)

In [15]:
#seleccion de variable independiente
X = df_unido.iloc[:,:-1].values
#seleccion de variable dependiente u objetivo
y=df_unido.iloc[:,83]

In [16]:
#cantidad de muestras por clase
print(f"{y.value_counts()}",f"{y.value_counts().sum()}" )

Normal        68424
DoS           52471
DDoS          48413
Probe         36372
BFA            1110
Web-Attack      192
BOTNET          164
Name: Label, dtype: int64 207146


# 2. Datos extraídos con cicflowmeter version Java

In [2]:
#diccionario de acceso a archivos de extraccion de flujos
dir_trafico = pl.Path("/ML-Tesis-Attacks/PcapToCsv/windows/trafico/")
"""crear diccionario para el acceso de imagenes"""
dicc_traficoJ = {
    'Normal':list(dir_trafico.glob('normal/*.csv')),
    'BOTNET': list(dir_trafico.glob('Botnet/*.csv')),
    'BFA': list(dir_trafico.glob('brute-force-attack/*.csv')),  
    'DDoS': list(dir_trafico.glob('DDoS/*.csv')),  
    'Dos': list(dir_trafico.glob('dos/**/*.csv')),  
    'Probe': list(dir_trafico.glob('Probe/*.csv')),  
    'Web-Attack': list(dir_trafico.glob('web_attack/*.csv'))  
}

In [3]:
#mostrar nro de archivos
listar_archivos(dicc_traficoJ)

Clase: Normal
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
Clase: BOTNET
1
Clase: BFA
1
2
3
4
Clase: DDoS
1
2
3
Clase: Dos
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
Clase: Probe
1
2
3
4
Clase: Web-Attack
1
2
3
4
5
6


In [4]:
#unir datos extraidos de la misma clase y etiquetarlos
etiquetar_unirdatosdeclase(dicc_traficoJ, 'Label', '/ML-Tesis-Attacks/datos_SDN/claseDatosJW/')

En proceso de clase Normal...
Proceso completado.... 
 Total de archivos unidos: 22
En proceso de clase BOTNET...
Proceso completado.... 
 Total de archivos unidos: 1
En proceso de clase BFA...
Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase DDoS...
Proceso completado.... 
 Total de archivos unidos: 3
En proceso de clase Dos...
Proceso completado.... 
 Total de archivos unidos: 18
En proceso de clase Probe...
Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase Web-Attack...
Proceso completado.... 
 Total de archivos unidos: 6


In [5]:
#importar
df_bfajw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/BFA.csv')
df_botnetjw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/BOTNET.csv')
df_ddosjw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/DDoS.csv')
df_dosjw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/Dos.csv')
df_normaljw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/Normal.csv')
df_probejw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/Probe.csv')
df_webattackjw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/Web-Attack.csv')

In [6]:
#concatenar dataset
df_jw = pd.concat([df_bfajw, df_botnetjw, df_ddosjw, df_dosjw, df_normaljw, df_probejw, df_webattackjw], axis=0)

In [7]:
#cantidad de muestras por clase
df_jw.Label.value_counts()

Dos           69651
Normal        60114
Probe         36723
BFA            1136
Web-Attack      204
BOTNET          164
DDoS             25
Name: Label, dtype: int64

# 3. Datos extraídos con cicflowmeter version Python

In [3]:
#diccionario de acceso a archivos de extraccion de flujos
dir_traficoP = pl.Path("/ML-Tesis-Attacks/PcapToCsv/ubuntu/trafico/")
"""crear diccionario para el acceso de imagenes"""
dicc_traficoP = {
    'Normal':list(dir_traficoP.glob('normal/**/*.csv')),
    'BOTNET': list(dir_traficoP.glob('Botnet/*.csv')),
    'BFA': list(dir_traficoP.glob('brute-force-attack/*.csv')),  
    'DDoS': list(dir_traficoP.glob('DDoS/*.csv')),  
    'Dos': list(dir_traficoP.glob('dos/**/*.csv')),  
    'Probe': list(dir_traficoP.glob('Probe/*.csv')),  
    'Web-Attack': list(dir_traficoP.glob('web_attack/*.csv'))  
}

In [4]:
listar_archivos(dicc_traficoP)

Clase: Normal
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
Clase: BOTNET
1
2
Clase: BFA
1
2
3
4
Clase: DDoS
1
2
3
4
Clase: Dos
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
Clase: Probe
1
2
3
4
Clase: Web-Attack
1
2
3
4
5
6


In [5]:
#unir datos por clase
etiquetar_unirdatosdeclase(dicc_traficoP,'Label', '/ML-Tesis-Attacks/datos_SDN/claseDatosPU/')

En proceso de clase Normal...
Proceso completado.... 
 Total de archivos unidos: 22
En proceso de clase BOTNET...
Proceso completado.... 
 Total de archivos unidos: 2
En proceso de clase BFA...
Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase DDoS...
Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase Dos...
Proceso completado.... 
 Total de archivos unidos: 18
En proceso de clase Probe...
Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase Web-Attack...
Proceso completado.... 
 Total de archivos unidos: 6


In [6]:
#importar archivos de datos generados
df_bfapu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/BFA.csv')
df_botnetpu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/BOTNET.csv')
df_ddospu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/DDoS.csv')
df_dospu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/Dos.csv')
df_normalpu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/Normal.csv')
df_probepu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/Probe.csv')
df_webattackpu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/Web-Attack.csv')

In [7]:
#concatenar dataset
df_pu = pd.concat([df_bfapu, df_botnetpu, df_ddospu, df_dospu, df_normalpu, df_probepu, df_webattackpu], axis=0)

In [8]:
#cantidad de muestras por clase
df_pu.Label.value_counts()

DDoS          398643
Normal         53784
Probe          41980
Dos            33842
BFA              582
Web-Attack       120
BOTNET            92
Name: Label, dtype: int64